### ISSUE TITLE
Symbol's status is void on copilot-login

### ISSUE THREAD

**Issue by scottanderson42** [zerolfx/copilot.el#79]
First-time installation,`copilot-login` gives the following error: `Symbol's value as variable is void: status`

Emacs 27.2, manual install of copilot.el:
```
(add-to-list 'load-path "/path-to-cloned-repo")
(require 'copilot)

(add-hook 'prog-mode-hook 'copilot-mode)
(with-eval-after-load 'company
  ;; disable inline previews
  (delq 'company-preview-if-just-one-frontend company-frontends))

(define-key copilot-completion-map (kbd "<tab>") 'copilot-accept-completion)
(define-key copilot-completion-map (kbd "TAB") 'copilot-accept-completion)
```

`*copilot stderr*` is empty.

`*copilot events*`:
```
server-reply (id:2) Tue Jan 10 17:18:36 2023:
(:jsonrpc "2.0" :id 2 :result "OK")

server-reply (id:3) ERROR Tue Jan 10 17:18:36 2023:
(:jsonrpc "2.0" :id 3 :error
          (:code 1000 :message "Not authenticated: NotSignedIn"))

client-request (id:4) Tue Jan 10 17:21:10 2023:
(:jsonrpc "2.0" :id 4 :method "signInInitiate" :params
          (:dummy "signInInitiate"))

server-reply (id:4) Tue Jan 10 17:21:11 2023:
(:jsonrpc "2.0" :id 4 :result
          (:status "PromptUserDeviceFlow" :userCode "166E-1CAB" :verificationUri "https://github.com/login/device" :expiresIn 899 :interval 5))
```

Stack trace:

```
Debugger entered--Lisp error: (void-variable status)
  (s-equals-p status "AlreadySignedIn")
  (if (s-equals-p status "AlreadySignedIn") (progn (user-error "Already signed in as %s" user)))
  (let* ((:user (plist-get input0 :status)) (user-code (plist-get input0 :userCode)) (verification-uri (plist-get input0 :verificationUri))) (if (s-equals-p status "AlreadySignedIn") (progn (user-error "Already signed in as %s" user))) (if (display-graphic-p) (progn (gui-set-selection 'CLIPBOARD user-code) (read-from-minibuffer (format "Your one-time code %s is copied. Press ENTER to op..." user-code)) (browse-url verification-uri) (read-from-minibuffer "Press ENTER if you finish authorizing.")) (read-from-minibuffer (format "First copy your one-time code: %s. Press ENTER to ..." user-code)) (read-from-minibuffer (format "Please open %s in your browser. Press ENTER if you..." verification-uri))) (message "Verifying...") (condition-case err (progn (if copilot--connection nil (copilot--start-agent)) (jsonrpc-request copilot--connection 'signInConfirm (list :userCode user-code))) (jsonrpc-error (message "Authentication failure: %s" (alist-get 'jsonrpc-error-message (cdr (cdr err)))))) (let ((input0 (progn (if copilot--connection nil (copilot--start-agent)) (jsonrpc-request copilot--connection 'checkStatus '(:dummy "checkStatus"))))) (let* nil (message "Authenticated as GitHub user %s." user))))
  (let ((input0 (progn (if copilot--connection nil (copilot--start-agent)) (jsonrpc-request copilot--connection 'signInInitiate '(:dummy "signInInitiate"))))) (let* ((:user (plist-get input0 :status)) (user-code (plist-get input0 :userCode)) (verification-uri (plist-get input0 :verificationUri))) (if (s-equals-p status "AlreadySignedIn") (progn (user-error "Already signed in as %s" user))) (if (display-graphic-p) (progn (gui-set-selection 'CLIPBOARD user-code) (read-from-minibuffer (format "Your one-time code %s is copied. Press ENTER to op..." user-code)) (browse-url verification-uri) (read-from-minibuffer "Press ENTER if you finish authorizing.")) (read-from-minibuffer (format "First copy your one-time code: %s. Press ENTER to ..." user-code)) (read-from-minibuffer (format "Please open %s in your browser. Press ENTER if you..." verification-uri))) (message "Verifying...") (condition-case err (progn (if copilot--connection nil (copilot--start-agent)) (jsonrpc-request copilot--connection 'signInConfirm (list :userCode user-code))) (jsonrpc-error (message "Authentication failure: %s" (alist-get 'jsonrpc-error-message (cdr (cdr err)))))) (let ((input0 (progn (if copilot--connection nil (copilot--start-agent)) (jsonrpc-request copilot--connection 'checkStatus '...)))) (let* nil (message "Authenticated as GitHub user %s." user)))))
  copilot-login()
  funcall-interactively(copilot-login)
  #<subr call-interactively>(copilot-login record nil)
  apply(#<subr call-interactively> copilot-login (record nil))
  call-interactively@ido-cr+-record-current-command(#<subr call-interactively> copilot-login record nil)
  apply(call-interactively@ido-cr+-record-current-command #<subr call-interactively> (copilot-login record nil))
  call-interactively(copilot-login record nil)
  command-execute(copilot-login record)
  execute-extended-command(nil "copilot-login")
  (unwind-protect (execute-extended-command current-prefix-arg chosen-item-name) (smex-rank chosen-item))
  (if smex-custom-action (let ((action smex-custom-action)) (setq smex-custom-action nil) (funcall action chosen-item)) (unwind-protect (execute-extended-command current-prefix-arg chosen-item-name) (smex-rank chosen-item)))
  (let* ((chosen-item-name (smex-completing-read commands initial-input)) (chosen-item (intern chosen-item-name))) (if smex-custom-action (let ((action smex-custom-action)) (setq smex-custom-action nil) (funcall action chosen-item)) (unwind-protect (execute-extended-command current-prefix-arg chosen-item-name) (smex-rank chosen-item))))
  smex-read-and-run(("toggle-debug-on-error" "copilot-login" "package-install" "about-emacs" "copilot-diagnose" "package-refresh-contents" "package-initialize" "search-forward-regexp" "magit-blame" "sort-lines" "capitalize-region" "load-file" "downcase-region" "compare-windows" "apply-macro-to-region-lines" "fill-region-as-paragraph" "flycheck-mode" "replace-regexp" "fill-region" "wdired-change-to-wdired-mode" "describe-function" "string-inflection-underscore" "describe-key" "auto-complete-mode" "shell" "upcase-region" "mc/insert-numbers" "query-replace" "js2-jsx-mode" "rgrep" "python-mode" "align-regexp" "web-mode" "string-inflection-camelcase" "transient-mark-mode" "comment-region" "undo-tree-redo" "coffee-mode" "tide-refactor" "javascript-mode" "tide-rename-symbol" "help" "set-variable" "magit-version" "json-pretty-print" "global-auto-complete-mode" "ert-results-jump-between-summary-and-result" "toggle-word-wrap" "highlight-indentation-mode" ...))
  (if (smex-already-running) (smex-update-and-rerun) (and smex-auto-update (smex-detect-new-commands) (smex-update)) (smex-read-and-run smex-ido-cache))
  smex()
  funcall-interactively(smex)
  #<subr call-interactively>(smex nil nil)
  apply(#<subr call-interactively> smex (nil nil))
  call-interactively@ido-cr+-record-current-command(#<subr call-interactively> smex nil nil)
  apply(call-interactively@ido-cr+-record-current-command #<subr call-interactively> (smex nil nil))
  call-interactively(smex nil nil)
  command-execute(smex)
```

Happy to provide anything else that helps.

**Comment by jakeisnt** [comment]
Having this problem on Emacs 28.2 as well! Not sure what's going wrong.

**Comment by zerolfx** [comment]
@scottanderson42 

Macro
```elisp
  (copilot--dbind
      (:status :user :userCode user-code :verificationUri verification-uri)
      (copilot--request 'signInInitiate '(:dummy "signInInitiate"))
```
expands to
```elisp
  (let
      ((input0
        (copilot--request 'signInInitiate
                          '(:dummy "signInInitiate"))))
    (let*
        ((status
          (plist-get input0 :status))
         (user
          (plist-get input0 :user))
         (user-code
          (plist-get input0 :userCode))
         (verification-uri
          (plist-get input0 :verificationUri)))
      (when
          (s-equals-p status "AlreadySignedIn")
        (user-error "Already signed in as %s" user))
```
on my machine.


But in your stack trace, it expands to `(let* ((:user (plist-get input0 :status)) (user-code (plist-get input0 :userCode))`, which is unexpected.

**Comment by zerolfx** [comment]
My suggestion is to check your `dash.el`'s version.

**Comment by scottanderson42** [comment]
@zerolfx I have dash 2.14.1 installed.

**Comment by scottanderson42** [comment]
2.19.1 fixed the issue, thanks!

**Comment by zerolfx** [comment]
Thanks for your confirmation. I'm just looking at `dash.el`'s changelog.

**Comment by coyotespike** [comment]
I am, sadly, getting this error.

- M1 Mac
- Doom Emacs
- `emacs-plus` 30.1

```
Debugger entered--Lisp error: (void-variable status)
  copilot-login()
  funcall-interactively(copilot-login)
  command-execute(copilot-login record)
  execute-extended-command(nil "copilot-login" nil)
  funcall-interactively(execute-extended-command nil "copilot-login" nil)
  command-execute(execute-extended-command)
```

Which is not a very helpful stack trace!

**Comment by coyotespike** [comment]
Okay, I put this in a scratch buffer and evaluated it:

```

(defun copilot-login ()
  "Login to Copilot."
  (interactive)
  (let* ((response (copilot--request 'signInInitiate '(:dummy "signInInitiate")))
         (status (plist-get response :status))
         (user (plist-get response :user))
         (user-code (plist-get response :userCode))
         (verification-uri (plist-get response :verificationUri)))
    
    (message "Full response: %S" response) ;; Add this for debugging
    
    (when (string-equal status "AlreadySignedIn")
      (user-error "Already signed in as %s" user))
    
    (if (display-graphic-p)
        (progn
          (gui-set-selection 'CLIPBOARD user-code)
          (read-from-minibuffer 
           (format "Your one-time code %s is copied. Press ENTER to open GitHub in your browser. 
                    If your browser does not open automatically, browse to %s." 
                   user-code verification-uri))
          (browse-url verification-uri)
          (read-from-minibuffer "Press ENTER if you finish authorizing."))
      (read-from-minibuffer 
       (format "First copy your one-time code: %s. Press ENTER to continue." user-code))
      (read-from-minibuffer 
       (format "Please open %s in your browser. Press ENTER if you finish authorizing." 
               verification-uri)))
    
    (copilot--log 'info "Verifying...")
    
    (condition-case err
        (copilot--request 'signInConfirm (list :userCode user-code))
      (jsonrpc-error 
       (user-error "Authentication failure: %s" 
                   (alist-get 'jsonrpc-error-message (cddr err)))))
    
    (let ((user (plist-get (copilot--request 'checkStatus '(:dummy "checkStatus")) :user)))
      (copilot--log 'info "Authenticated as GitHub user %s." user)))) 

```

And this worked! However, now in messages I get:

```
(Rocket.Chat.ReactNative-33a6108187) Starting tsserver...
(Rocket.Chat.ReactNative-33a6108187) tsserver server started successfully.
Error running timer: (wrong-type-argument number-or-marker-p nil) [16 times]
[jsonrpc] (warning) Sentinel for copilot server still hasn't run, deleting it!
[jsonrpc] Server exited with status 9
Copilot: Copilot server started. [2 times]
[jsonrpc] Server exited with status 1
Copilot: Copilot OK.
```

So this will require further debugging - no clue why `jsonrpc` is mad here.

**Comment by bbatsov** [comment]
@coyotespike See #396 - there was a regression that was just addressed in #397.

**Comment by coyotespike** [comment]
@bbatsov Oh, man - likely it worked on Spacemacs just due to the timing, I downloaded right as this fix went into production!

Any odds it now works in Doom! Many thanks for your work fixing it!!
